File: old-projects/vector-defence-silverlight-port/src/Particles/SlowingIndicator.ts
import { Sprite } from "../Sprite";
import { Game } from "../Game";
import { Util } from "../Util";

export class SlowingIndicator extends Sprite {
  private readonly target: Sprite;
  private readonly lineFrom: Sprite | null;
  private readonly fadeBy: number;
  private readonly color: string;
  private readonly fixedX: number;
  private readonly fixedY: number;

  constructor(x: number, y: number, targetSprite: Sprite);
  constructor(from: Sprite, targetSprite: Sprite);
  constructor(xOrFrom: number | Sprite, yOrTarget: number | Sprite, maybeTarget?: Sprite) {
    super();
    if (typeof xOrFrom === "number") {
      this.fixedX = xOrFrom;
      this.fixedY = yOrTarget as number;
      this.target = maybeTarget as Sprite;
      this.lineFrom = null;
      this.color = "#ffff00";
      this.alpha = 0.8;
      this.fadeBy = Util.calculateIntervalToComplete(1);
    } else {
      this.fixedX = 0;
      this.fixedY = 0;
      this.lineFrom = xOrFrom;
      this.target = yOrTarget as Sprite;
      this.color = "#008000";
      this.alpha = 0.7;
      this.fadeBy = 0;
    }
  }

  override animate(): void {
    if (this.target.markedForRemoval || (this.lineFrom !== null && this.lineFrom.markedForRemoval)) {
      this.alpha = 0;
    } else {
      this.alpha -= this.fadeBy * Game.currentFrameMultiplier;
    }
    if (this.alpha <= 0) {
      this.markedForRemoval = true;
    }
  }

  override render(context: CanvasRenderingContext2D): void {
    const x1 = this.lineFrom !== null ? this.lineFrom.x : this.fixedX;
    const y1 = this.lineFrom !== null ? this.lineFrom.y : this.fixedY;
    context.save();
    context.globalAlpha = Math.max(0, this.alpha);
    context.strokeStyle = this.color;
    context.lineWidth = 1;
    context.beginPath();
    context.moveTo(x1, y1);
    context.lineTo(this.target.x, this.target.y);
    context.stroke();
    context.restore();
  }
}

File: silverlight-ports/vector-defence/original-csharp-source/Particles/SlowingIndicator.cs
using System.Windows.Media;
using System.Windows.Shapes;

namespace VectorDefenceSL.Particles
{
    public class SlowingIndicator : Sprite
    {
        private readonly Sprite target;
        private readonly Brush stroke;
        private readonly Line line;
        private readonly Sprite lineFrom;
        private readonly double fadeBy;

        public SlowingIndicator(double x, double y, Sprite targetSprite) : this(targetSprite)
        {
            stroke = new SolidColorBrush(Colors.Yellow);
            line.Stroke = stroke;
            alpha = .8;
            line.X1 = x;
            line.Y1 = y;
            fadeBy = Util.CalculateIntervalToComplete(1);
        }

        public SlowingIndicator(Sprite from, Sprite targetSprite)
            : this(targetSprite)
        {
            stroke = new SolidColorBrush(Colors.Green);
            line.Stroke = stroke;
            alpha = .7;
            fadeBy = 0;
            lineFrom = from;
        }

        private SlowingIndicator(Sprite targetSprite)
        {
            target = targetSprite;
            mainElement = line = new Line
                                     {
                                         StrokeThickness = 1,
                                         Stroke = stroke,
                                         IsHitTestVisible = false,
                                     };
            Game.AddElement(mainElement);
        }

        public override void Animate()
        {
            if (target.MarkedForRemoval || (lineFrom != null && lineFrom.MarkedForRemoval))
            {
                alpha = 0;
            }
            else
            {
                alpha -= fadeBy * Game.CurrentFrameMultiplier;
            }

            if (alpha <= 0)
            {
                MarkedForRemoval = true;
            }
        }

        public override void Render()
        {
            if (lineFrom != null)
            {
                line.X1 = lineFrom.X;
                line.Y1 = lineFrom.Y;
            }
            line.X2 = target.X;
            line.Y2 = target.Y;
            stroke.Opacity = alpha;
        }
    }
}